### PR TITLE
Add OpInfo for binary ufuncs

### LINKED ISSUE #54294: Port the basic binary operations (add, sub, mul and div, true_divide) to use OpInfos
This issue is a more specific part of the overall OpInfo porting issue https://github.com/pytorch/pytorch/issues/54261.

This will require resolving why we currently have both OpInfos and method_tests() entries for div. We should ensure all historic cases covered by method_tests() are covered by the OpInfos' sample inputs, too.

cc @ezyang @gchanan @zou3519 @bdhirsh @jbschlosser @anjali411 @mruberry @VitalyFedyunin @walterddr

### PR DESCRIPTION
**Pull request by pmeier**
Closes #54294, supersedes #54323. This PR is a superset of #56227 and (partially) #55201. Cc @imaginary-person 

---

This adds a `class BinaryUfuncInfo(OpInfo)` to simplify testing of binary ufuncs. As a starting point this also adds numeric reference tests for `add`, `sub`, `mul`, `div`, `true_divide`, and `floor_divide`.

### CODE REVIEW

File: torch/testing/_internal/common_methods_invocations.py [hunk 1]
```diff
@@ -348,7 +348,7 @@ class UnaryUfuncInfo(OpInfo):
       - they typically support the out kwarg
       - they typically have NumPy or SciPy references
     See NumPy's universal function documentation
-    (https://numpy.org/doc/1.18/reference/ufuncs.html) for more details
+    (https://numpy.org/doc/stable/reference/ufuncs.html) for more details
```
Comment (pmeier) [inline torch/testing/_internal/common_methods_invocations.py:351]: Driveby that was approved in the original PR https://github.com/pytorch/pytorch/pull/54323/files#r599260687

File: test/run_test.py [hunk 2]
```diff
@@ -264,7 +264,7 @@
     'test_jit',
     'test_jit_profiling',
     'test_torch',
-    'test_binary_ufuncs'
+    'test_binary_ufuncs',
```
Comment (pmeier) [inline test/run_test.py:267]: Without this comma, the strings are concatenated to `test_binary_ufuncstest_numpy_interop`.

File: test/test_binary_ufuncs.py [hunk 3]
```diff
@@ -1183,6 +1184,34 @@ def test_binary_ops_with_scalars(self, device):
...
+            np_dtype = np.float16
+
+            def to_numpy(t):
+                return np.array(t.tolist(), dtype=np_dtype)
+
+        else:
+            np_dtype = torch_to_numpy_dtype_dict[dtype]
+
+            def to_numpy(t):
+                return np.array(t.cpu(), dtype=np_dtype)
```
Comment (pmeier) [inline test/test_binary_ufuncs.py:1203]: If we need this pattern in multiple tests, we should factor it out.

File: test/test_binary_ufuncs.py [hunk 4]
```diff
@@ -1183,6 +1184,34 @@ def test_binary_ops_with_scalars(self, device):
...
+
+        # TODO: turn these into separate tests
+        for sample in op.sample_inputs(device, dtype):
+            torch_inputs = (sample.input,) + sample.args
+            numpy_inputs = [to_numpy(t) for t in torch_inputs]
+
+            torch_output = getattr(torch, op.name)(*torch_inputs).cpu()
+            numpy_output = torch.as_tensor(op.ref(*numpy_inputs), dtype=torch_output.dtype)
+
+            assert_close(torch_output, numpy_output, equal_nan=True)
```
Comment (pmeier) [inline test/test_binary_ufuncs.py:1213]: 1. This blocks merging until the newly added functions of `torch.testing` are green-lighted.
2. `assert_close()` currently does not support complex inputs. Thus this PR is further blocked until this is included.

File: test/run_test.py (hunk 2 above)
Comment (mruberry) [reply test/run_test.py:267]: yikes, nice fix

File: test/test_binary_ufuncs.py [hunk 5]
```diff
@@ -1183,6 +1183,34 @@ def test_binary_ops_with_scalars(self, device):
...
                         self.assertEqual(expected, torch_op(first, second))
 
+    # Tests that the function and its (array-accepting) reference produce the same
+    #   values on a range of tensors, including empty tensors, scalar tensors,
+    #   1D tensors and a large 2D tensor
+    @onlyOnCPUAndCUDA
+    @ops(binary_ufuncs)
+    def test_reference_numerics(self, device, dtype, op):
+        if dtype == torch.bfloat16:
+            np_dtype = np.float16
```
Comment (mruberry) [inline test/test_binary_ufuncs.py:1193]: np.float32 may be more appropriate since it and bfloat16 have the same dynamic range

File: test/test_binary_ufuncs.py [hunk 6]
```diff
@@ -1183,6 +1183,34 @@ def test_binary_ops_with_scalars(self, device):
...
+
+            def to_numpy(t):
+                return np.array(t.tolist(), dtype=np_dtype)
```
Comment (mruberry) [inline test/test_binary_ufuncs.py:1196]: torch tensors can be converted to NumPy arrays using .numpy():

```
a = t.cpu().numpy().astype(np_dtype)
```

File: test/test_binary_ufuncs.py (hunk 3 above)
Comment (mruberry) [reply test/test_binary_ufuncs.py:1202]: See above comment

File: test/test_binary_ufuncs.py [hunk 7]
```diff
@@ -1183,6 +1183,34 @@ def test_binary_ops_with_scalars(self, device):
...
+
+            def to_numpy(t):
+                return np.array(t.cpu(), dtype=np_dtype)
+
+        # TODO: turn these into separate tests
+        for sample in op.sample_inputs(device, dtype):
+            torch_inputs = (sample.input,) + sample.args
+            numpy_inputs = [to_numpy(t) for t in torch_inputs]
+
+            torch_output = getattr(torch, op.name)(*torch_inputs).cpu()
```
Comment (mruberry) [inline test/test_binary_ufuncs.py:1209]: if you want to call the function variant of an op then `op` is a callable and will call the function variant:

```
torch_output = op(*torch_inputs).cpu()
```

File: test/test_binary_ufuncs.py [hunk 8]
```diff
@@ -1183,6 +1183,34 @@ def test_binary_ops_with_scalars(self, device):
...
+            numpy_output = torch.as_tensor(op.ref(*numpy_inputs), dtype=torch_output.dtype)
```
Comment (mruberry) [inline test/test_binary_ufuncs.py:1210]: torch.from_numpy()

File: test/test_binary_ufuncs.py (hunk 4 above)
Comment (mruberry) [reply test/test_binary_ufuncs.py:1212]: I would just use assertEqual for now

File: torch/testing/_internal/common_methods_invocations.py [hunk 9]
```diff
@@ -400,6 +400,67 @@ def __init__(self,
         #   outside a function's domain.
         self._domain_eps = 1e-5
 
+
+def sample_inputs_binary(op_info, device, dtype, requires_grad):
```
Comment (mruberry) [inline torch/testing/_internal/common_methods_invocations.py:404]: See https://github.com/pytorch/pytorch/pull/56227, where @imaginary-person is added sample inputs for binary pwise ops, too.

Maybe we can take the union of the both of them?

File: torch/testing/_internal/common_methods_invocations.py [hunk 10]
```diff
@@ -400,6 +400,67 @@ def __init__(self,
...
+                 *,
+                 ref,  # a reference function
+                 domain=(None, None),  # the [low, high) domain of the function
+                 dtypes=all_types_and_complex_and(torch.half, torch.bfloat16),
+                 sample_inputs_func=sample_inputs_binary,
+                 **kwargs):
+        super(BinaryUfuncInfo, self).__init__(name,
+                                              dtypes=dtypes,
+                                              sample_inputs_func=sample_inputs_func,
+                                              supports_inplace_autograd=False,
```
Comment (mruberry) [inline torch/testing/_internal/common_methods_invocations.py:453]: What happens with this approach if the constructor is used with `supports_inplace_autograd=True`? Won't the OpInfo constructor see two of the same argument?

File: torch/testing/_internal/common_methods_invocations.py [hunk 11]
```diff
@@ -5380,8 +5441,36 @@ def gradcheck_wrapper_triangular_input(op, input, *args, upper=False, **kwargs):
                    safe_casts_outputs=True),
 ]
 
+op_db += [
```
Comment (mruberry) [inline torch/testing/_internal/common_methods_invocations.py:5444]: Put these in the regular op_db (at least for now)

File: torch/testing/_internal/common_methods_invocations.py [hunk 12]
```diff
@@ -5380,8 +5441,36 @@ def gradcheck_wrapper_triangular_input(op, input, *args, upper=False, **kwargs):
...
+    BinaryUfuncInfo('add',
+                    ref=np.add,
+                    dtypes=all_types_and_complex_and(torch.bool, torch.half, torch.bfloat16),
+                    supports_out=True,
+                    ),
+    BinaryUfuncInfo('mul',
```
Comment (mruberry) [inline torch/testing/_internal/common_methods_invocations.py:5450]: see https://github.com/pytorch/pytorch/pull/56227, which adds mul and sub

File: torch/testing/_internal/common_methods_invocations.py [hunk 13]
```diff
@@ -5380,8 +5441,36 @@ def gradcheck_wrapper_triangular_input(op, input, *args, upper=False, **kwargs):
                    safe_casts_outputs=True),
 ]
 
+op_db += [
+    BinaryUfuncInfo('add',
```
Comment (mruberry) [inline torch/testing/_internal/common_methods_invocations.py:5445]: I think we already have opinfos for add and mul?

File: torch/testing/_internal/common_methods_invocations.py [hunk 14]
```diff
@@ -400,6 +400,67 @@ def __init__(self,
...
+                    args=(make_tensor(s2, device, dtype,
+                                      low=low, high=high,
+                                      requires_grad=requires_grad),))
+        for s1, s2 in test_cases)
+
+
+# Metadata class for binary "universal functions (ufuncs)" that accept two
+# tensor and have common properties
+class BinaryUfuncInfo(OpInfo):
+    """Operator information for 'universal binary functions (binary ufuncs).'
```
Comment (mruberry) [inline torch/testing/_internal/common_methods_invocations.py:430]: Nice description

File: test/test_binary_ufuncs.py [hunk 15]
```diff
@@ -1183,6 +1183,34 @@ def test_binary_ops_with_scalars(self, device):
                         self.assertEqual(expected, python_op(first, second))
                         self.assertEqual(expected, torch_op(first, second))
 
+    # Tests that the function and its (array-accepting) reference produce the same
+    #   values on a range of tensors, including empty tensors, scalar tensors,
+    #   1D tensors and a large 2D tensor
+    @onlyOnCPUAndCUDA
+    @ops(binary_ufuncs)
+    def test_reference_numerics(self, device, dtype, op):
```
Comment (mruberry) [inline test/test_binary_ufuncs.py:1191]: I would call this "test_reference_numerics" because those tests in test_unary_ufuncs.py generate their own massive inputs; this is more like "test_reference_samples"?

What do you think is the future for a test like this? Is it worth comparing both with samples and large random tensors? If so, should we also compare the unary pwise ops vs their references using samples? If so, should we put this in test_ops.py and filter it to the binary and unary ufuncs?

File: test/test_binary_ufuncs.py (hunk 8 above)
Comment (pmeier) [reply test/test_binary_ufuncs.py:1210]: [`torch.from_numpy`](https://pytorch.org/docs/stable/generated/torch.from_numpy.html) only works if the input is a `np.ndarray`. Depending on the inputs the `numpy` op might also return a scalar:

```python
import numpy as np

print(type(np.add(False, True)))
print(type(np.add(0, 1)))
print(type(np.add(0.0, 1.0)))
```

```
<class 'numpy.bool_'>
<class 'numpy.int64'>
<class 'numpy.float64'>
```

File: test/test_binary_ufuncs.py (hunk 6 above)
Comment (pmeier) [reply test/test_binary_ufuncs.py:1196]: This will fail for `bfloat16` since `numpy` does not know this dtype.

File: torch/testing/_internal/common_methods_invocations.py (hunk 13 above)
Comment (pmeier) [reply torch/testing/_internal/common_methods_invocations.py:5445]: `add` yes, `mul` no.

File: test/test_binary_ufuncs.py (hunk 15 above)
Comment (pmeier) [reply test/test_binary_ufuncs.py:1191]: I have no answer to this yet. @mattip thoughts?

File: torch/testing/_internal/common_methods_invocations.py (hunk 9 above)
Comment (pmeier) [reply torch/testing/_internal/common_methods_invocations.py:404]: They are already. The only thing missing here is the explicit use of a python scalar instead of a scalar tensor.

File: test/test_binary_ufuncs.py (hunk 6 above)
Comment (mruberry) [reply test/test_binary_ufuncs.py]: Good point; I suppose it will have to map the NumPy dtype to a torch dtype and perform the conversion before converting to NumPy

**Comment by pmeier** [comment]
Blocked by #57119.

**Comment by imaginary-person** [comment]
Hello @pmeier, is it possible to test the `alpha` kwarg for `add` & `sub`, as @mruberry [suggested](https://github.com/pytorch/pytorch/pull/56227#discussion_r618918304)? Thank you!

**Comment by pmeier** [comment]
> Hello @pmeier, is it possible to test the `alpha` kwarg for `add` & `sub`, as @mruberry [suggested](https://github.com/pytorch/pytorch/pull/56227#discussion_r618918304)? Thank you!

Hey @imaginary-person. Yes, this is possible and reasonable to do. Thanks for letting me know. IMO the easiest way is to add a separate `BinaryUfuncInfo` since [`numpy.add`](https://numpy.org/doc/1.20/reference/generated/numpy.add) doesn't support this keyword.

**Comment by pmeier** [comment]
@imaginary-person I've added two additional OpInfo's for the `alpha` kwarg. Could you have a look?

File: torch/testing/_internal/common_methods_invocations.py [hunk 16]
```diff
@@ -3716,30 +3800,59 @@ def gradcheck_wrapper_triangular_input(op, input, *args, upper=False, **kwargs):
...
+        aliases=("divide",),
+        ref=lambda input, other, **kwargs: np.true_divide(input, other),
+        # prevent zero division
+        domain=(1.0, None),
+        sample_inputs_func=partial(sample_inputs_binary, rounding_mode=None),
+        dtypes=all_types_and_complex_and(torch.bool, torch.half, torch.bfloat16),
+        supports_out=True,
+        skips=(
+            SkipInfo("TestCommon", "test_variant_consistency_jit", dtypes=[torch.float32, torch.complex64]),
+        )
```
Comment (pmeier) [inline torch/testing/_internal/common_methods_invocations.py:3825]: For some reason this test seem to fail if the `rounding_mode` is explicitly set or `floor_divide` (L410) or `true_divide` (L5521) are used directly. They all fail because the exported / imported dtype does not match the original. This happens for `float32` for all of them and for `complex64` for this OpInfo only.

Since all our dtypes pass, I'm guessing this is not a test error but a legitimate failure.

**Comment by imaginary-person** [comment]
> @imaginary-person I've added two additional OpInfo's for the `alpha` kwarg. Could you have a look?

Hello @pmeier, thank you for adding additional OpInfos of `add` & `sub` for the `alpha` kwarg. However, @mruberry had recommended testing with only one sample input for the `alpha` kwarg, so can you please do so, if possible? Thanks!

**Comment by pmeier** [comment]
Superseded by #61964.
